core/state_transition.go

```diff
@@ -258,50 +264,76 @@ func (st *StateTransition) to() common.Address {
 
 // payFees deducts gas and gateway fees from sender balance and adds the purchased amount of gas to the state.
 func (st *StateTransition) payFees(eHardFork bool) error {
-	if !currency.IsWhitelisted(st.vmRunner, st.msg.FeeCurrency()) {
+	var isWhiteListed bool
+	if eHardFork {
+		isWhiteListed = st.sysCtx.IsWhitelisted(st.msg.FeeCurrency())
+	} else {
+		isWhiteListed = currency.IsWhitelisted(st.vmRunner, st.msg.FeeCurrency())
+	}
+	if !isWhiteListed {
 		log.Trace("Fee currency not whitelisted", "fee currency address", st.msg.FeeCurrency())
 		return ErrNonWhitelistedFeeCurrency
 	}
 
-	feeVal := new(big.Int).Mul(new(big.Int).SetUint64(st.msg.Gas()), st.gasPrice)
-
+	effectiveFee := new(big.Int).Mul(new(big.Int).SetUint64(st.msg.Gas()), st.gasPrice)
 	// If GatewayFeeRecipient is unspecified, the gateway fee value is ignore and the sender is not charged.
 	if st.msg.GatewayFeeRecipient() != nil {
-		feeVal.Add(feeVal, st.msg.GatewayFee())
+		effectiveFee.Add(effectiveFee, st.msg.GatewayFee())
 	}
-
-	if !st.canPayFee(st.msg.From(), feeVal, st.msg.FeeCurrency(), eHardFork) {
-		return ErrInsufficientFundsForFees
+	if err := st.canPayFee(st.msg.From(), effectiveFee, st.msg.FeeCurrency(), eHardFork); err != nil {
+		return err
 	}
 	if err := st.gp.SubGas(st.msg.Gas()); err != nil {
 		return err
 	}
 
 	st.initialGas = st.msg.Gas()
 	st.gas += st.msg.Gas()
-	err := st.debitFee(st.msg.From(), feeVal, st.msg.FeeCurrency())
+	err := st.debitFee(st.msg.From(), effectiveFee, st.msg.FeeCurrency())
 	return err
 }
 
-func (st *StateTransition) canPayFee(accountOwner common.Address, fee *big.Int, feeCurrency *common.Address, eHardfork bool) bool {
+// canPayFee checks whether accountOwner's balance can cover transaction fee.
+//
+// For native token(CELO) as feeCurrency:
+//   - Pre-Espresso: it ensures balance >= GasPrice * gas + gatewayFee
+//   - Post-Espresso: it ensures balance >= GasFeeCap * gas + value + gatewayFee
+// For non-native tokens(cUSD, cEUR, ...) as feeCurrency:
+//   - Pre-Espresso: it ensures balance > GasPrice * gas + gatewayFee
+//   - Post-Espresso: it ensures balance >= GasFeeCap * gas + gatewayFee
+func (st *StateTransition) canPayFee(accountOwner common.Address, fee *big.Int, feeCurrency *common.Address, espresso bool) error {
```

Comment:
`fee` -> `effectiveFee`? All the shadowing here makes it a little confusing what `fee` is at any given moment.